File: app/src/test/java/com/example/myapplication/core/today/TodayResolverTest.kt
package com.example.myapplication.core.today

import org.junit.Assert.assertEquals
import org.junit.Test

class TodayResolverTest {
    @Test
    fun inactiveGoalReturnsNoGoal() {
        assertEquals(TodayResult.NoGoal, TodayResolver.resolve(false, emptyList(), 100L))
    }

    @Test
    fun activeGoalWithEveryWorkoutCompletedReturnsGoalComplete() {
        val sessions = listOf(ScheduledWorkout("done", 0, 90L, completed = true))

        assertEquals(TodayResult.GoalComplete, TodayResolver.resolve(true, sessions, 100L))
    }

    @Test
    fun earliestIncompleteWorkoutBySequenceWinsEvenWhenInputIsUnorderedAndOverdue() {
        val earliest = ScheduledWorkout("first", 1, 90L, completed = false)
        val sessions = listOf(
            ScheduledWorkout("later", 2, 95L, completed = false),
            ScheduledWorkout("completed", 0, 80L, completed = true),
            earliest,
        )

        assertEquals(TodayResult.Workout(earliest), TodayResolver.resolve(true, sessions, 100L))
    }

    @Test
    fun futureEarliestWorkoutReturnsRecoveryWithNextDueDate() {
        val next = ScheduledWorkout("next", 0, 105L, completed = false)

        assertEquals(TodayResult.Recovery(105L), TodayResolver.resolve(true, listOf(next), 100L))
    }
}

File: flutter/test/core/today/today_resolver_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_app/core/today/today_resolver.dart';

void main() {
  test('inactiveGoalReturnsNoGoal', () {
    expect(
      TodayResolver.resolve(hasActiveGoal: false, workouts: [], todayEpochDay: 100),
      TodayResultNoGoal(),
    );
  });

  test('activeGoalWithEveryWorkoutCompletedReturnsGoalComplete', () {
    final sessions = [ScheduledWorkout(id: 'done', sequence: 0, dueEpochDay: 90, completed: true)];

    expect(
      TodayResolver.resolve(hasActiveGoal: true, workouts: sessions, todayEpochDay: 100),
      TodayResultGoalComplete(),
    );
  });

  test('earliestIncompleteWorkoutBySequenceWinsEvenWhenInputIsUnorderedAndOverdue', () {
    final earliest = ScheduledWorkout(id: 'first', sequence: 1, dueEpochDay: 90, completed: false);
    final sessions = [
      ScheduledWorkout(id: 'later', sequence: 2, dueEpochDay: 95, completed: false),
      ScheduledWorkout(id: 'completed', sequence: 0, dueEpochDay: 80, completed: true),
      earliest,
    ];

    expect(
      TodayResolver.resolve(hasActiveGoal: true, workouts: sessions, todayEpochDay: 100),
      TodayResultWorkout(earliest),
    );
  });

  test('futureEarliestWorkoutReturnsRecoveryWithNextDueDate', () {
    final next = ScheduledWorkout(id: 'next', sequence: 0, dueEpochDay: 105, completed: false);

    expect(
      TodayResolver.resolve(hasActiveGoal: true, workouts: [next], todayEpochDay: 100),
      TodayResultRecovery(105),
    );
  });
}
